src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -48,413 +46,1245 @@ internal static class ThreadPoolGlobals
-                            // This way we don't need to rearrange the items already in the queue; they'll be found
-                            // correctly exactly where they are.  One subtlety here is that we need to make sure that
-                            // if head is currently < tail, it remains that way.  This happens to just fall out from
-                            // the bit-masking, because we only do this if tail == int.MaxValue, meaning that all
-                            // bits are set, so all of the bits we're keeping will also be set.  Thus it's impossible
-                            // for the head to end up > than the tail, since you can't set any more bits than all of
-                            // them.
-                            //
-                            m_headIndex &= m_mask;
-                            m_tailIndex = tail = m_tailIndex & m_mask;
-                            Debug.Assert(m_headIndex <= m_tailIndex);
+                            // Attempt to acquire the slot for Dequeuing.
+                            if (Interlocked.CompareExchange(ref _queueEnds.Dequeue, position + 1, position) == position)
+                            {
+                                var item = slot.Item;
+                                slot.Item = null;
+
+                                // make the slot appear empty in the next generation
+                                Volatile.Write(ref slot.SequenceNumber, position + 1 + _slotsMask);
+                                return item;
+                            }
+                        }
+                        else if (sequenceNumber - position < Full)
+                        {
+                            // The sequence number was less than what we needed, which means we cannot return this item.
+                            // Check if we have reached Enqueue and return null indicating the segment is in empty state.
+                            // NB: reading stale _frozenForEnqueues is fine - we would just spin once more
+                            var currentEnqueue = Volatile.Read(ref _queueEnds.Enqueue);
+                            if (currentEnqueue == position || (_frozenForEnqueues && currentEnqueue == position + FreezeOffset))
+                            {
+                                return null;
+                            }
+
+                            // The enqueuer went ahead and took a slot, but it has not finished filling the value.
+                            // We cannot return `null` since the segment is not empty, so we must retry.
+                            // We should be careful though to not starve the enqueuer that we are waiting for.
+                            // In a worst case it may have lower priority than us, so we have to sleep(1) at some point.
+                            // Let the SpinWait handle that.
+                            spinner.SpinOnce();
```

Comment:
`Sleep(0)` is sufficient, I have seen that allowing a `Sleep(1)` here significantly increases global work item latency under some types of contention that hit this path.
```suggestion
                            spinner.SpinOnce(sleep1Threshold: -1);
```